### ISSUE TITLE
Adaptive lighting does not adjust brightness of Lightener group correctly

### ISSUE THREAD

**Issue by Patrick1610** [basnijholt/adaptive-lighting#978]
# Home Assistant Adaptive Lighting Issue Template

## Bug Reports

If you need help with using or configuring Adaptive Lighting, please [open a Q&A discussion thread here](https://github.com/basnijholt/adaptive-lighting/discussions/new?category=q-a) instead.

### Before submitting a bug report, please follow these troubleshooting steps:

Please confirm that you have completed the following steps:

- [ ] I have updated to the [latest Adaptive Lighting version](https://github.com/basnijholt/adaptive-lighting/releases) available in [HACS](https://hacs.xyz/).
- [ ] I have reviewed the [Troubleshooting Section](https://github.com/basnijholt/adaptive-lighting#troubleshooting) in the [README](https://github.com/basnijholt/adaptive-lighting#readme).
- [ ] (If using Zigbee2MQTT) I have read the [Zigbee2MQTT troubleshooting guide](https://github.com/basnijholt/adaptive-lighting#zigbee2mqtt) in the [README](https://github.com/basnijholt/adaptive-lighting#readme).
- [ ] I have checked the [V2 Roadmap](https://github.com/basnijholt/adaptive-lighting/discussions/291) and [open issues](https://github.com/basnijholt/adaptive-lighting/issues) to ensure my issue isn't a duplicate.

### Required information for bug reports:

Please include the following information in your issue.

*Issues missing this information may not be addressed.*

1.  **Debug logs** captured while the issue occurred. [See here for instructions on enabling debug logging](https://github.com/basnijholt/adaptive-lighting#troubleshooting):

```

```

2.  [Your Adaptive Lighting configuration](https://github.com/basnijholt/adaptive-lighting#gear-configuration):

Just added 1 of multiple lightener groups, left the rest of the settings default

```

3.  (If using Zigbee2MQTT), provide your configuration files (**remove all personal information before posting**):
    - `devices.yaml`
    - `groups.yaml`
    - `configuration.yaml` ⚠️; **Warning** _**REMOVE ALL of the PERSONAL INFORMATION BELOW before posting**_ ⚠️;
      - mqtt: `server`:
      - mqtt: `user`:
      - mqtt: `password`:
      - advanced: `pan_id`:
      - advanced: `network_key`:
      - anything in `log_syslog` if you use this
    - Brand and model number of problematic light(s)
```

```

4.  Describe the bug and how to reproduce it:
There is a form of light groups from another custom integration, called Lightener. It allows you to specify different brightness percentages per light in the group. So when the group is set to 50%, one light can be 10% and another 70% for example. 

When I set this group as the destination for AL, it does not work. 

AL should set the lightener to a percentage (for example 50% and the underlying lights should be adapter by lightener. 
Instead, AL sets all underlying lights to 50% and the Lightener is not adapter to 50% at all. 

I can create screenshots or a short screen recording later if needed. 

5. Steps to reproduce the behavior:
Create a lightener
Set only lightener as light for AL
Turn on Lightener light

**Comment by daniel1v** [comment]
I am experiencing the same issue. It seems like adaptive lighting expands light groups and then applies its brightness and color temperature to the single lights instead of the group itself. this is fine for standard light groups but leads to unexpected behavior in lightener groups, because of the non linear brightness mapping. As lightener inherits from light group, it is recognized as a light group.

the current behavior was a bit confusing for me, because one could add a light group but also single lights to adaptive lighting. this gives the impression, that adaptive lighting is able to control the group light itself and not  only the contained single lights.

@basnijholt, can you confirm this? maybe there could be a switch to turn off group expansion?

**Comment by Patrick1610** [comment]
@daniel1v, I didn't understand the problem as deep as you do, so thanks for clarifying!

It would be an amazing solution if this could be turned on/off in the configuration of each AL instance. I hope this is possible @basnijholt 🙏🏼

**Comment by Patrick1610** [comment]
In addition: manual control is also looking at the individual lights. It would be nice if that keeps working as is.

**Comment by basnijholt** [comment]
There could be an option to turn off the expansion but this would break the manual control.

**Comment by Patrick1610** [comment]
It would be great if that was possible, because the use of Lightener makes the relativity between lights possible, which makes the automatic lighting of AL near perfect!

Wouldn't it be possible to have an additional option to switch between manual control monitoring for the group or the individual lights? 
For example: That the lights are loaded twice into the integration and used without expansion for the control-part and used with expansion for the monitoring-part of the instance?

**Comment by daniel1v** [comment]
> There could be an option to turn off the expansion but this would break the manual control.

yeah, it is right, that the manual control for single group members could be impacted, when the group is not expanded anymore. one could argue, that this is totally ok, when selecting this imaginary option. 

To overcome the issue, I have several suggestions in my mind, but imho the best would be:

by default, AL should control the group entity itself, and not expand light groups for adaption. 

for the detection of manual control, AL should watch the group entity and also all the group members:
- as soon as manual control is detected for the group entity itself, AL should stop adapting the group entity
- as soon as manual control  is detected for a group member, AL should switch to the same behavior as now: do not control the group entity, instead expand the group and exclude manually controlled single light entities

I think this solution would have no negative side effects for the existing users and could also be the default behavior of AL but also could only be available via config option.

what do you think @basnijholt ?

**Comment by Patrick1610** [comment]
@daniel1v I agree with you that it would be great to see a solution that allows us to keep using manual control. However, only having the option to disable group expansion (at first) would be most important for me, since that would allow me to use lightener and set up relative lighting scenarios. 

In the case that manual control will be fixed I only disagree with the following:

`as soon as manual control is detected for a group member, AL should switch to the same behavior as now: do not control the group entity, instead expand the group and exclude manually controlled single light entities `

Let me elaborate. When using Lightener the following could happen. Let's say AL is at 50%, and lightener would give us the following scenario:
- Light A: 10%
- Light B: 90%
- Light C: 70% 

When I manually set Light C to 80%, and that disables manual control for all lights, the next time AL updates Light A and B will see massive changes, since they both will be changed to 50%. I think when a light is changed that is a part of a group (and that group is in AL), and the imaginary 'Group expansion' button is disabled, the group should be marked as manually controlled as a whole. 
Or there should be an extra option so users can choose between those two versions of manual control in combination with disabling group expansion.

**Comment by daniel1v** [comment]
@Patrick1610 The proposed solution was intended to be more general and, of course, the manual control detection should not trigger, when a light state is changed by the group. In fact, there may be some details, I do not know about. 

To make a clear statement: I would also support a feature request to simply have an option to turn off group expansion everywhere and live with the drawbacks.

@basnijholt would it be possible to get such a feature? Should a separate feature request be opened? I would really appreciate using AL and lightener together.

**Comment by Patrick1610** [comment]
@basnijhol Just wondering and trying to understand, because the tool and its capabilities are amazing! But how does AL work for you? Are all the lights in your house/room at the same level, or do you use a lot of scenes that disable AL during the day? 

I can't imagine having all lights at the same level, since I have lights of different brands or dimmer with classic bulbs that are much brighter than other bulbs, but that might be very specific for me. 

Wouldn't it be possible to have a feature like Lightener inside AL? Where you can specify the percentages of the lights after you have chosen the lights. So AL still calculates a percentage throughout the day, but what that percentage means per light is adjustable. Just check the lightener repo to see what I mean 😉

**Comment by c10l** [comment]
I have a slightly different use case. I have a few areas that are illuminated by multiple lights. This is true for areas with 3-bulb chandeliers, areas with multiple spotlights and even light strips.

Generally I don't have an issue with differences between lights in a group, but even at 1% if all lights are on at the same time, that's a lot more brightness than what I want at certain times of the day.

With Lightener I can control the general brightness of the area, so if Adaptive Lighting can control the Lightener group instead of the individual lights, that means each light group always behaves predictably.

I have very little use for per-light control within a group, except for testing purposes.

**Comment by tylerhartwig** [comment]
@basnijholt would you be open to a PR for this issue? If so -- what would be the desired functionality? it seems there's multiple ways to go about solving this one.

**Comment by iggyha** [comment]
I just started using Lightener with Adaptive Lighting and I was surprised to see all of the lights go down together - it’s almost a perfect pair. If this could be updated, it would really make a lot of sense at least for me and how my brain works.

**Comment by Benthejunebug** [comment]
I have come up with a solution that works pretty darn well.

First you have to make a virtual light via [HASS Virtual](https://github.com/twrecked/hass-virtual/blob/master/README.md) 

-- if using HAOS you may need to make the virtual.yaml manually with the File Editor Addon in the root/folder where your configuration.yaml is but don't change the '/config/virtual.yaml' when you make an instance of the integration. Thanks to this comment/thread [virtual.yaml will not be visible · Issue #140 · twrecked/hass-virtual](https://github.com/twrecked/hass-virtual/issues/140#issuecomment-2546418124) for helping me figure out how to get around that issue

Put this virtual light you make in AL settings in place of your Lightener Light.

Then you use this blueprint [Linked Entities](https://community.home-assistant.io/t/linked-entities-keep-mutlple-entities-state-in-sync-lights-switches-etc/662836) to sync the lights (I tried HACS Sync Groups but that didn't work).

Took some fiddling but it works! there can be a bit of a delay at moments when turned on but a small delay here and there doesn't really matter to me.

Happy Adaptive Lightenering!

**Comment by tylerhartwig** [comment]
> I have come up with a solution that works pretty darn well.
> 
> First you have to make a virtual light via [HASS Virtual](https://github.com/twrecked/hass-virtual/blob/master/README.md)
> 
> -- if using HAOS you may need to make the virtual.yaml manually with the File Editor Addon in the root/folder where your configuration.yaml is but don't change the '/config/virtual.yaml' when you make an instance of the integration. Thanks to this comment/thread [virtual.yaml will not be visible · Issue #140 · twrecked/hass-virtual](https://github.com/twrecked/hass-virtual/issues/140#issuecomment-2546418124) for helping me figure out how to get around that issue
> 
> Put this virtual light you make in AL settings in place of your Lightener Light.
> 
> Then you use this blueprint [Linked Entities](https://community.home-assistant.io/t/linked-entities-keep-mutlple-entities-state-in-sync-lights-switches-etc/662836) to sync the lights (I tried HACS Sync Groups but that didn't work).
> 
> Took some fiddling but it works! there can be a bit of a delay at moments when turned on but a small delay here and there doesn't really matter to me.
> 
> Happy Adaptive Lightenering!

I tried this, the problem is manual control doesn't get recognized because of the virtual component

**Comment by redvinzent** [comment]
Just came here to +1 this.
If breaking the manual control is an issue, this could be an opt-it checkbox with a warning. Myself I don't care about manual control of single lights, but the ability to use lightener for, say, a three-bulb bathroom light and turn off two bulbs at night would be perfect. 
I saw that there's a possibility to work around this, but I (maybe ignorantly, apologies then) think that adding an option to choose the target >group instead of >all lights in a group should be fairly straightforward?

**Comment by Ultra9k** [comment]
I still can't believe it but Grok just solved this compatibility issue for me in a very simple way by editing the light.py of the lightener integration:

It added this line in the __init__ method of the LightenerLight class:
`self._attr_extra_state_attributes = {"is_hue_group": True}`

And "reinforced it in async_write_ha_state to make sure it always stays", whatever that means:
```
current_attrs["is_hue_group"] = True
self._attr_extra_state_attributes = current_attrs
```

Apparently the Adaptive Lighting code checks this specific condition before expanding a group/light into its individual members:
```
if "entity_id" in state.attributes and not state.attributes.get("is_hue_group", False):
    # expand the group
```

By forcing "is_hue_group": True on the Lightener virtual light, Adaptive Lighting sees it as a "Hue-style group" and skips the expansion step completely.

It works! I hope this helps someone. Here you have the modified light.py

[light.py](https://github.com/user-attachments/files/25376385/light.py)

**Comment by tylerhartwig** [comment]
@Ultra9k Does this allow for manual control outside of Home Assistant as well?

**Comment by Ultra9k** [comment]
> [@Ultra9k](https://github.com/Ultra9k) Does this allow for manual control outside of Home Assistant as well?

Yes, I’m still amazed by how well everything works with this little change.

Edit: apparently each individual light of the group can’t then be manually controlled. Well, it can but after a minute it will adopt the lightener group settings again. So it isn’t perfect but it still works for me.

**Comment by basnijholt** [comment]
This is covered by v1.32.0. Target the Lightener group in your Adaptive Lighting profile and turn **Expand light groups** off (`expand_light_groups: false`).

I tested this in Home Assistant using Lightener 2.4.0's actual implementation. A 50% group target correctly produced 10% and 70% for the members, preserving the mapping.

Manual control then applies to the group as a whole. Changing one member does not pause just that member, so later group updates can overwrite it.

Closing as resolved by the existing option. If it still fails on your devices with this setting, please share current diagnostics and logs.
